astropy/coordinates/builtin_frames/icrs_cirs_transforms.py

```diff
@@ -288,3 +289,15 @@ def hcrs_to_hcrs(from_coo, to_frame):
     else:
         # like CIRS, we do this self-transform via ICRS
         return from_coo.transform_to(ICRS()).transform_to(to_frame)
+
+
+@frame_transform_graph.transform(FunctionTransformWithFiniteDifference, TETE, TETE)
+def tete_to_tete(from_coo, to_frame):
+    if (np.all(from_coo.obstime == to_frame.obstime) and
+            np.all(from_coo.obsgeoloc == to_frame.obsgeoloc) and
+            np.all(from_coo.obsgeovel == to_frame.obsgeovel)):
+        return to_frame.realize_frame(from_coo.data)
+    else:
+        # can't do self-transform via ITRS due to infinite recursion
```

Comment:
Why is this? Bug in the code or physical reason? Would be good to add explanation (and perhaps TODO if just a bug), especially as the self-transform is used in `tete_to_itrs`

Reply:
This is due to a flaw in how the TETE<->ITRS transformations are written (see https://github.com/astropy/astropy/pull/10909#issuecomment-714213513 for the same problem with the TEME<->ITRS transformations).  The TETE<->ITRS transformations should be rewritten to not rely on an existing TETE<->TETE loopback, and then the loopback can be written to depend on the TETE<->ITRS transformations without triggering infinite recursion.

Reply:
In short, an `obstime` change needs to be handled in those transformations as an ITRS loopback rather than a TETE loopback.

Reply:
@ayshih I actually think the way it's done here is preferable. Your re-written TEM[T]E<->ITRS transformations **still** require a trip through ICRS, it's just not explicit anymore.

This is because an ITRS<->ITRS loopback for different ```obstimes``` goes back through CIRS, which in turn goes back through ICRS. So the **actual transform** called still goes through CIRS and ICRS, you've just added a step in the lookup chain.

In the existing scenario:

1) Code sees that it needs a TETE->TETE' self transform
2) That transform goes through ICRS
3) Code looks up TETE->ICRS transform
4) That transform is TETE->CIRS->ICRS.
5) Code performs TETE->CIRS->ICRS-CIRS'-TETE' 

If I implemented the changes in your PR:

1) Code sees that it needs a TETE->TETE' self transform
2) That transform goes through ITRS
3) Code looks up TETE->ITRS transform
4) The TETE->ITRS involves an ITRS self-transform
5) The ITRS->ITRS' transform is ITRS->CIRS->ICRS->CIRS'->ITRS'
6) Code does TETE->ITRS->CIRS->ICRS->CIRS'->ITRS'->TETE'

I don't see the second scenario is better. As equatorial frames, TEME and TETE are brothers of CIRS and shouldn't sit "downstream" of ITRS in the transform chain.

Reply:
Hi @mhvk - I've added a hopefully clearer comment to the self-transform.

I've also added a "needs-discussion" label to **prevent anyone merging** even though it is approved and passing.

The reason is that I have a branch implementing improved accuracy in CIRS by way of adding a topocentric position to that frame. Given the similarities between CIRS and TETE, the intermediate rotation transforms in that branch are similar to the ones here.

However, although that branch passes almost all tests, there is some odd behaviour in ITRS->CIRS->ALTAZ that I don't currently understand. It's making me nervous there's a flaw in the transform I haven't seen.

Reply:
> In the existing scenario:
> 
> 1. Code sees that it needs a TETE->TETE' self transform
> 2. That transform goes through ICRS
> 3. Code looks up TETE->ICRS transform
> 4. That transform is TETE->CIRS->ICRS.
> 5. Code performs TETE->CIRS->ICRS->CIRS'->TETE'

I believe your comparison is misleading because you don't have defined TETE<->CIRS transformations.  So, it's actually:

4. That transform is TETE->GCRS->CIRS->ICRS.
5. Code performs TETE->GCRS->CIRS->ICRS->CIRS'->GCRS'->TETE'

That is, the difference between your current implementation and my suggested re-jigger is that the intermediate frame is ITRS instead of GCRS.  Importantly, the fact that the TETE->ICRS composite transformation goes through GCRS instead of through ITRS is *luck*: they traverse the same number of edges in the graph, so it comes down to the order they happen to be stored in the graph.

My perspective is that you should decide which frame you want to consider as "upstream" – GCRS or ITRS – and handle a changing `obstime` there, rather than handling it in the "downstream" frame.

> As an aside, I hate the way that the transform graph shows, for example, an ITRS->ITRS' loopback, since it's a lie! There's an argument for removing all of these shortcut functions and just making everything Earth-related go through ICRS if we don't write real self-transforms...

It's been suggested before (#8996).  If you're worried about hidden transformations, it pervades beyond just the loopbacks.  For example, the TETE->ITRS transformation is a single edge in the graph, but internally it will handle a change in `obstime` by a loop through ICRS using at least four other transformations and two other frames.

Reply:
@StuartLittlefair, @ayshih - thanks for the explanation, makes sense! And the comment now helps.

On the larger picture, I agree it is misleading to have self-transformations shown if they do not in fact exists. If we were to remove all self-transformations, would the graph automatically do the right thing, i.e., go to the nearest frame that does have a self-transformation and go back? Independently, we should remove the arrow...  Anyway, probably worth a separate issue, as we're not likely to change it for 4.2!

Reply:
> On the larger picture, I agree it is misleading to have self-transformations shown if they do not in fact exists.

Self-transformations have to be implemented at some level to show up on the graph.  The issue here is the how to implement the loopback if a "real" self-transform doesn't exist (i.e., one that is just calculations rather than relying on the existing transformation graph.)

> If we were to remove all self-transformations, would the graph automatically do the right thing, i.e., go to the nearest frame that does have a self-transformation and go back?

No.  Unless #8996 is addressed, the default behavior is that the transform will see the same frame as source and destination and perform a no-op.

Reply:
Yes - what @ayshih says.

Thanks for correcting my example, although really it's TETE->GCRS->ICRS->GCRS'->TETE. Which is conceptually correct, as shown in this figure from P6 of the "Sofa tools for earth attitude doc"

<img width="678" alt="Screenshot 2020-10-22 at 16 27 00" src="https://user-images.githubusercontent.com/4570807/96894319-8172fc80-1483-11eb-8078-fe3162f90076.png">

> My perspective is that you should decide which frame you want to consider as "upstream" – GCRS or ITRS – and handle a changing obstime as a loopback there, rather than handling it in the "downstream" frame (TETE).

I *think* this is how I've done it here, no? In that the self transform goes through ICRS, which is really TETE->GCRS->ICRS. Have I missed your meaning?

Reply:
Whoops, you're correct.  I misread the graph.  You're correct.  TETE->ICRS goes through GCRS instead of through ITRS->CIRS.

Reply:
We both did, and I wrote the code. I really hate the graph.

Reply:
Yes, at least the no-op transform should only happen if the frames are equal...